sw/device/lib/dif/dif_uart.c

```diff
@@ -0,0 +1,409 @@
+    return false;
+  }
+
+  // Set baudrate, enable RX and TX, configure parity
+  uint32_t reg = (nco_masked << UART_CTRL_NCO_OFFSET);
+  reg |= (1u << UART_CTRL_TX);
+  reg |= (1u << UART_CTRL_RX);
+  if (config->parity_enable == kDifUartEnable) {
+    reg |= (1u << UART_CTRL_PARITY_EN);
+  }
+  if (config->parity == kDifUartParityOdd) {
+    reg |= (1u << UART_CTRL_PARITY_ODD);
+  }
+  reg32_write(uart->base_addr, UART_CTRL(0), reg);
+
+  // Reset RX/TX FIFOs
+  reg = (1u << UART_FIFO_CTRL_RXRST);
+  reg |= (1u << UART_FIFO_CTRL_TXRST);
+  reg32_write(uart->base_addr, UART_FIFO_CTRL(0), reg);
+
+  // Disable interrupts
+  reg32_write(uart->base_addr, UART_INTR_ENABLE(0), 0u);
+
+  return true;
+}
+
+/**
+ * Write up to |bytes_requested| number of bytes to the TX FIFO.
+ */
+static size_t uart_bytes_send(const dif_uart_t *uart, const uint8_t *data,
+                              size_t bytes_requested) {
+  size_t bytes_written = 0;
+  while ((bytes_written < bytes_requested) && !uart_tx_full(uart)) {
+    uart_tx_fifo_write(uart, data[bytes_written]);
+    ++bytes_written;
+  }
+
+  return bytes_written;
+}
+
+/**
+ * Read up to |bytes_requested| number of bytes from the RX FIFO.
+ */
+static size_t uart_bytes_receive(const dif_uart_t *uart, size_t bytes_requested,
+                                 uint8_t *data) {
+  size_t bytes_read = 0;
+  while ((bytes_read < bytes_requested) && !uart_rx_empty(uart)) {
+    data[bytes_read] = uart_rx_fifo_read(uart);
+    ++bytes_read;
+  }
+
+  return bytes_read;
+}
+
+bool dif_uart_init(reg32_t base_addr, const dif_uart_config_t *config,
+                   dif_uart_t *uart) {
+  if (uart == NULL || config == NULL || base_addr.inner_ptr == 0) {
+    return false;
+  }
+
+  uart->base_addr = base_addr;
+
+  return uart_configure(uart, config);
+}
+
+bool dif_uart_configure(const dif_uart_t *uart,
+                        const dif_uart_config_t *config) {
+  if ((uart == NULL) || (config == NULL)) {
+    return false;
+  }
+
+  return uart_configure(uart, config);
+}
+
+bool dif_uart_watermark_rx_set(const dif_uart_t *uart,
+                               dif_uart_watermark_t watermark) {
+  if (uart == NULL) {
+    return false;
+  }
+
+  // Check if the requested watermark is valid, and get a corresponding
+  // register definition to be written
+  uint32_t rxi_level;
+  switch (watermark) {
+    case kDifUartWatermarkByte1:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL1;
+      break;
+    case kDifUartWatermarkByte4:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL4;
+      break;
+    case kDifUartWatermarkByte8:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL8;
+      break;
+    case kDifUartWatermarkByte16:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL16;
+      break;
+    case kDifUartWatermarkByte30:
+      rxi_level = UART_FIFO_CTRL_RXILVL_RXLVL30;
+      break;
+    default:
+      return false;
+  }
+
+  // Set watermark level
+  uint32_t mask = rxi_level & UART_FIFO_CTRL_RXILVL_MASK;
```

Comment:
A minor point, but I don't think you need to `&` with `UART_FIFO_CTRL_RXILVL_MASK` here since all `..._RXLVLN` defines are within this mask by construction, right? You can pass `rxi_level` directly to `reg32_nonatomic_set_mask`.